pkg/instancetype/controller/vm/status.go

```diff
@@ -29,36 +30,51 @@ import (
 	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
 )
 
-func (h *revisionHandler) patchVM(
-	instancetypeStatusRef, preferenceStatusRef *virtv1.InstancetypeStatusRef,
+func getOriginalStatusRef(vm *virtv1.VirtualMachine) (
+	origInstancetypeStatus *virtv1.InstancetypeStatusRef,
+	origPreferenceStatus *virtv1.InstancetypeStatusRef,
+) {
+	if vm.Status.InstancetypeRef != nil {
+		origInstancetypeStatus = vm.Status.InstancetypeRef.DeepCopy()
+	}
+	if vm.Status.PreferenceRef != nil {
+		origPreferenceStatus = vm.Status.PreferenceRef.DeepCopy()
+	}
+	return origInstancetypeStatus, origPreferenceStatus
+}
+
+func (c *controller) patchStatus(
 	vm *virtv1.VirtualMachine,
+	origInstancetypeStatus, origPreferenceStatus *virtv1.InstancetypeStatusRef,
 ) error {
-	// Batch any writes to the VirtualMachine into a single PatchStatus() call to avoid races in the controller.
 	logger := func() *log.FilteredLogger { return log.Log.Object(vm) }
-	revisionPatch, err := GeneratePatch(instancetypeStatusRef, preferenceStatusRef)
+	revisionPatch, err := c.generateStatusPatch(vm, origInstancetypeStatus, origPreferenceStatus)
 	if err != nil || len(revisionPatch) == 0 {
 		return err
 	}
-	if _, err := h.virtClient.VirtualMachine(vm.Namespace).PatchStatus(
+	if _, err := c.clientset.VirtualMachine(vm.Namespace).PatchStatus(
 		context.Background(), vm.Name, types.JSONPatchType, revisionPatch, metav1.PatchOptions{},
 	); err != nil {
-		logger().Reason(err).Error("Failed to update VirtualMachine with instancetype and preference ControllerRevision references.")
+		logger().Reason(err).Error("Failed to update VirtualMachine with instancetype and preference references.")
 		return err
 	}
 	return nil
 }
 
-func GeneratePatch(instancetypeStatusRef, preferenceStatusRef *virtv1.InstancetypeStatusRef) ([]byte, error) {
+func (c *controller) generateStatusPatch(
+	vm *virtv1.VirtualMachine,
+	origInstancetypeStatus, origPreferenceStatus *virtv1.InstancetypeStatusRef,
+) ([]byte, error) {
 	patchSet := patch.New()
-	if instancetypeStatusRef != nil {
+	if !equality.Semantic.DeepEqual(vm.Status.InstancetypeRef, origInstancetypeStatus) {
 		patchSet.AddOption(
-			patch.WithAdd("/status/instancetypeRef", instancetypeStatusRef),
+			patch.WithAdd("/status/instancetypeRef", vm.Status.InstancetypeRef),
```

Comment:
Use `patch.WithReplace` and add `patch.WithTest` for `origInstancetypeStatus` if `vm.Status.InstancetypeRef != nil`, otherwise use `patch.WithAdd`?